File: backend/app/schemas/alerts_v3.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field

from app.pydantic_compat import PYDANTIC_V2, ConfigDict


class AlertVariableDef(BaseModel):
    name: str = Field(..., min_length=1)
    # MVP: allow either free-form DSL or structured kind/params
    # (compiler supports both).
    dsl: Optional[str] = None
    kind: Optional[str] = None
    params: Optional[Dict[str, Any]] = None


class AlertDefinitionCreate(BaseModel):
    name: str = Field(..., min_length=1)
    target_kind: str = Field(..., min_length=1)
    target_ref: str = Field(..., min_length=1)
    exchange: Optional[str] = None

    evaluation_cadence: Optional[str] = None
    variables: List[AlertVariableDef] = Field(default_factory=list)
    condition_dsl: str = Field(..., min_length=1)

    trigger_mode: str = Field(default="ONCE_PER_BAR")
    throttle_seconds: Optional[int] = Field(default=None, ge=1)
    only_market_hours: bool = False
    expires_at: Optional[datetime] = None
    enabled: bool = True


class AlertDefinitionUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1)
    target_kind: Optional[str] = None
    target_ref: Optional[str] = None
    exchange: Optional[str] = None

    evaluation_cadence: Optional[str] = None
    variables: Optional[List[AlertVariableDef]] = None
    condition_dsl: Optional[str] = None

    trigger_mode: Optional[str] = None
    throttle_seconds: Optional[int] = Field(default=None, ge=1)
    only_market_hours: Optional[bool] = None
    expires_at: Optional[datetime] = None
    enabled: Optional[bool] = None


class AlertDefinitionRead(BaseModel):
    id: int
    name: str
    target_kind: str
    target_ref: str
    exchange: Optional[str] = None

    evaluation_cadence: str
    variables: List[AlertVariableDef] = Field(default_factory=list)
    condition_dsl: str

    trigger_mode: str
    throttle_seconds: Optional[int] = None
    only_market_hours: bool
    expires_at: Optional[datetime] = None
    enabled: bool

    last_evaluated_at: Optional[datetime] = None
    last_triggered_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    if PYDANTIC_V2:
        model_config = ConfigDict(from_attributes=True)
    else:  # pragma: no cover

        class Config:
            orm_mode = True


class CustomIndicatorCreate(BaseModel):
    name: str = Field(..., min_length=1)
    description: Optional[str] = None
    params: List[str] = Field(default_factory=list)
    body_dsl: str = Field(..., min_length=1)
    enabled: bool = True


class CustomIndicatorUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1)
    description: Optional[str] = None
    params: Optional[List[str]] = None
    body_dsl: Optional[str] = None
    enabled: Optional[bool] = None


class CustomIndicatorRead(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    params: List[str] = Field(default_factory=list)
    body_dsl: str
    enabled: bool
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    if PYDANTIC_V2:
        model_config = ConfigDict(from_attributes=True)
    else:  # pragma: no cover

        class Config:
            orm_mode = True


class AlertEventRead(BaseModel):
    id: int
    alert_definition_id: int
    symbol: str
    exchange: Optional[str] = None
    evaluation_cadence: Optional[str] = None
    reason: Optional[str] = None
    snapshot: Dict[str, Any] = Field(default_factory=dict)
    triggered_at: datetime
    bar_time: Optional[datetime] = None

    if PYDANTIC_V2:
        model_config = ConfigDict(from_attributes=True)
    else:  # pragma: no cover

        class Config:
            orm_mode = True


class AlertV3TestRequest(BaseModel):
    target_kind: str = Field(..., min_length=1)
    target_ref: str = Field(..., min_length=1)
    exchange: Optional[str] = None

    evaluation_cadence: Optional[str] = None
    variables: List[AlertVariableDef] = Field(default_factory=list)
    condition_dsl: str = Field(..., min_length=1)


class AlertV3TestResult(BaseModel):
    symbol: str
    exchange: str
    matched: bool
    bar_time: Optional[datetime] = None
    snapshot: Dict[str, Any] = Field(default_factory=dict)
    error: Optional[str] = None


class AlertV3TestResponse(BaseModel):
    evaluation_cadence: str
    results: List[AlertV3TestResult] = Field(default_factory=list)


__all__ = [
    "AlertVariableDef",
    "AlertDefinitionCreate",
    "AlertDefinitionUpdate",
    "AlertDefinitionRead",
    "CustomIndicatorCreate",
    "CustomIndicatorUpdate",
    "CustomIndicatorRead",
    "AlertEventRead",
    "AlertV3TestRequest",
    "AlertV3TestResult",
    "AlertV3TestResponse",
]

File: frontend/src/services/alertsV3.ts
export type AlertVariableDef = {
  name: string
  dsl?: string | null
  kind?: string | null
  params?: Record<string, unknown> | null
}

export type AlertDefinition = {
  id: number
  name: string
  broker_name: string
  target_kind: string
  target_ref: string
  symbol?: string | null
  exchange?: string | null
  action_type: 'ALERT_ONLY' | 'BUY' | 'SELL'
  action_params: Record<string, unknown>
  evaluation_cadence: string
  variables: AlertVariableDef[]
  condition_dsl: string
  trigger_mode: 'ONCE' | 'ONCE_PER_BAR' | 'EVERY_TIME'
  throttle_seconds?: number | null
  only_market_hours: boolean
  expires_at?: string | null
  enabled: boolean
  last_evaluated_at?: string | null
  last_triggered_at?: string | null
  created_at?: string | null
  updated_at?: string | null
}

export type AlertDefinitionCreate = {
  name: string
  broker_name?: string
  target_kind: string
  target_ref?: string | null
  symbol?: string | null
  exchange?: string | null
  action_type: 'ALERT_ONLY' | 'BUY' | 'SELL'
  action_params: Record<string, unknown>
  evaluation_cadence?: string | null
  variables: AlertVariableDef[]
  condition_dsl: string
  trigger_mode: 'ONCE' | 'ONCE_PER_BAR' | 'EVERY_TIME'
  throttle_seconds?: number | null
  only_market_hours: boolean
  expires_at?: string | null
  enabled: boolean
}

export type AlertDefinitionUpdate = Partial<AlertDefinitionCreate>

export type CustomIndicator = {
  id: number
  name: string
  description?: string | null
  params: string[]
  body_dsl: string
  enabled: boolean
  created_at?: string | null
  updated_at?: string | null
}

export type CustomIndicatorCreate = Omit<CustomIndicator, 'id' | 'created_at' | 'updated_at'>
export type CustomIndicatorUpdate = Partial<CustomIndicatorCreate>

export type AlertEvent = {
  id: number
  alert_definition_id: number
  symbol: string
  exchange?: string | null
  evaluation_cadence?: string | null
  reason?: string | null
  snapshot: Record<string, unknown>
  triggered_at: string
  bar_time?: string | null
}

export type AlertV3TestRequest = {
  broker_name?: string
  target_kind: string
  target_ref?: string | null
  symbol?: string | null
  exchange?: string | null
  evaluation_cadence?: string | null
  variables: AlertVariableDef[]
  condition_dsl: string
}

export type AlertV3TestResult = {
  symbol: string
  exchange: string
  matched: boolean
  bar_time?: string | null
  snapshot: Record<string, unknown>
  error?: string | null
}

export type AlertV3TestResponse = {
  evaluation_cadence: string
  results: AlertV3TestResult[]
}

async function _parseError(res: Response): Promise<string> {
  const text = await res.text()
  try {
    const parsed = JSON.parse(text) as { detail?: unknown }
    if (typeof parsed.detail === 'string') return parsed.detail
  } catch {
    // ignore
  }
  return text
}

export async function listAlertDefinitions(): Promise<AlertDefinition[]> {
  const res = await fetch('/api/alerts-v3/')
  if (!res.ok) {
    const detail = await _parseError(res)
    throw new Error(`Failed to load alerts (${res.status})${detail ? `: ${detail}` : ''}`)
  }
  return (await res.json()) as AlertDefinition[]
}

export async function createAlertDefinition(
  payload: AlertDefinitionCreate,
): Promise<AlertDefinition> {
  const res = await fetch('/api/alerts-v3/', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const detail = await _parseError(res)
    throw new Error(`Failed to create alert (${res.status})${detail ? `: ${detail}` : ''}`)
  }
  return (await res.json()) as AlertDefinition
}

export async function updateAlertDefinition(
  id: number,
  payload: AlertDefinitionUpdate,
): Promise<AlertDefinition> {
  const res = await fetch(`/api/alerts-v3/${id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const detail = await _parseError(res)
    throw new Error(`Failed to update alert (${res.status})${detail ? `: ${detail}` : ''}`)
  }
  return (await res.json()) as AlertDefinition
}

export async function deleteAlertDefinition(id: number): Promise<void> {
  const res = await fetch(`/api/alerts-v3/${id}`, { method: 'DELETE' })
  if (!res.ok) {
    const detail = await _parseError(res)
    throw new Error(`Failed to delete alert (${res.status})${detail ? `: ${detail}` : ''}`)
  }
}

export async function listCustomIndicators(): Promise<CustomIndicator[]> {
  const res = await fetch('/api/alerts-v3/indicators/')
  if (!res.ok) {
    const detail = await _parseError(res)
    throw new Error(
      `Failed to load custom indicators (${res.status})${detail ? `: ${detail}` : ''}`,
    )
  }
  return (await res.json()) as CustomIndicator[]
}

export async function createCustomIndicator(
  payload: CustomIndicatorCreate,
): Promise<CustomIndicator> {
  const res = await fetch('/api/alerts-v3/indicators/', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const detail = await _parseError(res)
    throw new Error(
      `Failed to create custom indicator (${res.status})${detail ? `: ${detail}` : ''}`,
    )
  }
  return (await res.json()) as CustomIndicator
}

export async function updateCustomIndicator(
  id: number,
  payload: CustomIndicatorUpdate,
): Promise<CustomIndicator> {
  const res = await fetch(`/api/alerts-v3/indicators/${id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const detail = await _parseError(res)
    throw new Error(
      `Failed to update custom indicator (${res.status})${detail ? `: ${detail}` : ''}`,
    )
  }
  return (await res.json()) as CustomIndicator
}

export async function deleteCustomIndicator(id: number): Promise<void> {
  const res = await fetch(`/api/alerts-v3/indicators/${id}`, { method: 'DELETE' })
  if (!res.ok) {
    const detail = await _parseError(res)
    throw new Error(
      `Failed to delete custom indicator (${res.status})${detail ? `: ${detail}` : ''}`,
    )
  }
}

export async function listAlertEvents(params?: {
  alertId?: number
  symbol?: string
  limit?: number
}): Promise<AlertEvent[]> {
  const url = new URL('/api/alerts-v3/events/', window.location.origin)
  if (params?.alertId != null) url.searchParams.set('alert_id', String(params.alertId))
  if (params?.symbol) url.searchParams.set('symbol', params.symbol)
  if (params?.limit != null) url.searchParams.set('limit', String(params.limit))

  const res = await fetch(url.toString())
  if (!res.ok) {
    const detail = await _parseError(res)
    throw new Error(`Failed to load events (${res.status})${detail ? `: ${detail}` : ''}`)
  }
  return (await res.json()) as AlertEvent[]
}

export async function testAlertExpression(
  payload: AlertV3TestRequest,
  options?: { limit?: number },
): Promise<AlertV3TestResponse> {
  const url = new URL('/api/alerts-v3/test', window.location.origin)
  if (options?.limit != null) url.searchParams.set('limit', String(options.limit))

  const res = await fetch(url.toString(), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const detail = await _parseError(res)
    throw new Error(`Failed to test expression (${res.status})${detail ? `: ${detail}` : ''}`)
  }
  return (await res.json()) as AlertV3TestResponse
}
